src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -298,12 +298,23 @@ impl Step for Cargo {
 
     /// Runs `cargo test` for `cargo` packaged with Rust.
     fn run(self, builder: &Builder<'_>) {
-        if self.stage < 2 {
-            eprintln!("WARNING: cargo tests on stage {} may not behave well.", self.stage);
+        // If stage is explicitly set or not lower than 2, keep it. Otherwise, make sure it's at least 2
+        // as tests for this step don't work with a lower stage.
+        let stage = if builder.config.explicit_stage_from_cli
```

Comment:
I would perhaps move this into `make_run`, to only apply this logic when the user explicitly asks for `src/tools/cargo` on the command line, and not when someone invokes the tests from the bootstrap code (in which case it might be a bit too magic that the stage changes). But no strong opinion on this.